src/search/parser/decorated_abstract_syntax_tree.cc

```diff
@@ -0,0 +1,563 @@
+    cout << indent << "LIST:" << endl;
+    indent = "| " + indent;
+    for (const DecoratedASTNodePtr &element : elements) {
+        element->dump(indent);
+    }
+}
+
+VariableNode::VariableNode(const string &name)
+    : name(name) {
+}
+
+plugins::Any VariableNode::construct(ConstructContext &context) const {
+    context.push_layer("Looking up variable '" + name + "'");
+    if (!context.has_variable(name)) {
+        context.construction_error("Variable '" + name + "' is not defined.");
+    }
+    context.pop_layer();
+    return context.get_variable(name);
+}
+
+void VariableNode::dump(string indent) const {
+    cout << indent << "VAR: " << name << endl;
+}
+
+BoolLiteralNode::BoolLiteralNode(const string &value)
+    : value(value) {
+}
+
+plugins::Any BoolLiteralNode::construct(ConstructContext &context) const {
+    context.push_layer("Constructing bool value from '" + value + "'");
+    istringstream stream(value);
+    bool x;
+    if ((stream >> boolalpha >> x).fail()) {
+        ABORT("Could not parse bool constant '" + value + "'"
+              " (this should have been caught before constructing this node).");
+    }
+    context.pop_layer();
+    return x;
+}
+
+void BoolLiteralNode::dump(string indent) const {
+    cout << indent << "BOOL: " << value << endl;
+}
+
+IntLiteralNode::IntLiteralNode(const string &value)
+    : value(value) {
+}
+
+plugins::Any IntLiteralNode::construct(ConstructContext &context) const {
+    context.push_layer("Constructing int value from '" + value + "'");
+    if (value.empty()) {
+        ABORT("Empty value in int constant '" + value + "'"
+              " (this should have been caught before constructing this node).");
+    } else if (value == "infinity") {
+        context.pop_layer();
+        return numeric_limits<int>::max();
+    }
+
+    char suffix = value.back();
+    string prefix = value;
+    int factor = 1;
+    if (isalpha(suffix)) {
+        suffix = static_cast<char>(tolower(suffix));
+        if (suffix == 'k') {
+            factor = 1000;
+        } else if (suffix == 'm') {
+            factor = 1000000;
+        } else if (suffix == 'g') {
+            factor = 1000000000;
+        } else {
+            ABORT("Invalid suffix in int constant '" + value + "'"
+                  " (this should have been caught before constructing this node).");
+        }
+        prefix.pop_back();
+    }
+
+    istringstream stream(prefix);
+    int x;
+    stream >> noskipws >> x;
+    if (stream.fail() || !stream.eof()) {
+        ABORT("Could not parse int constant '" + value + "'"
+              " (this should have been caught before constructing this node).");
+    }
+
+    int min_int = numeric_limits<int>::min();
+    // Reserve highest value for "infinity".
+    int max_int = numeric_limits<int>::max() - 1;
+    if (!utils::is_product_within_limits(x, factor, min_int, max_int)) {
+        context.construction_error("Integer constant too large: '" + value + "'");
```

Comment:
Technically it could also be too small right? But I guess we don't have negative Integers as options anywhere.

Reply:
I changed it to "Absolute value of integer constant too large". I considered "out of bounds" but since we also have bounds on parameters and they are checked elsewhere, I found this confusing.